### DISCUSSION TITLE
`add_exception_handler` type annotation does not accept coroutines

### DISCUSSION THREAD

**Discussion by fpiche** [encode/starlette#2399] [Potential Issue]
A related discussion was answered here but does not fully describe my issue https://github.com/encode/starlette/discussions/2391

Since the introduction of the `ExceptionHandler` type, a coroutine cannot be passed to `add_exception_handler`.

Example:

I have an async function with signature
```
async def exc_handler(self, request: Request, exc: HTTPException) -> JSONResponse: 
 ```
and am passing it to `add_exception_handler` like so: 
```
app.add_exception_handler(HTTPException, exc_handler)
```
But mypy is complaining that I’m passing the wrong type:
```
Argument 2 to "add_exception_handler" of "Starlette" has incompatible type "Callable[[Request, HTTPException], Coroutine[Any, Any, JSONResponse]]"; expected "Callable[[Request, Exception], Response | Awaitable[Response]] | Callable[[WebSocket, Exception], Awaitable[None]]"
```

A workaround i found for now is to add exception handlers during app creation since I'm using FastAPI: `FastAPI(exception_handlers={HTTPException: exc_handler})` since this accepts the coroutine type.

**Comment by Jodagito** [comment]
I'm going thought this exact same problem. Thing is when adding multiple exception handlers to the FastAPI object then it won't be clean at all, one could use the `exception_handlers` list from the FastAPI object and replace the `add_exception_handler` method, this new method would just populate the list of dicts.
